snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,271 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+from threading import Event, Lock
+from typing import Callable, Generic, Mapping, MutableMapping, Optional, Set, Sequence
+
+from snuba.utils.concurrent import execute
+from snuba.utils.streams.consumers.backends.abstract import ConsumerBackend
+from snuba.utils.streams.consumers.types import Message, TOffset, TStream, TValue
+
+
+ConsumerGroup = str
+
+
+@dataclass
+class CommitData(Generic[TStream, TOffset]):
+    consumer_group: ConsumerGroup
+    stream: TStream
+    offset: TOffset
+
+
+# TODO: This should be moved into the abstract consumer backend.
+class Topic(ABC, Generic[TStream]):
+    @abstractmethod
+    def __contains__(self, stream: TStream) -> bool:
+        raise NotImplementedError
+
+
+@dataclass
+class Subscription(Generic[TStream, TOffset]):
+    topics: Sequence[Topic[TStream]]
+    remote_consumer_group_offsets: MutableMapping[
+        TStream, MutableMapping[ConsumerGroup, TOffset]
+    ] = field(default_factory=lambda: defaultdict(dict))
+
+
+# TODO: TOffset needs to be bound as some sort of comparable type.
+
+
+class SynchronizedConsumerBackend(ConsumerBackend[TStream, TOffset, TValue]):
+    def __init__(
+        self,
+        backend: ConsumerBackend[TStream, TOffset, TValue],
+        commit_log_consumer_backend: ConsumerBackend[
+            # NOTE: These TStream and TOffset values for the commit log
+            # consumer do not need to match, but this simplifies things a bit
+            # so that we don't have to declare two additional type variables.
+            TStream,
+            TOffset,
+            CommitData[TStream, TOffset],
+        ],
+        remote_consumer_groups: Set[ConsumerGroup],
+        get_commit_log_streams_for_topic: Callable[[Topic[TStream]], Sequence[TStream]],
+    ) -> None:
+        self.__backend = backend
+        self.__commit_log_consumer_backend = commit_log_consumer_backend
+        self.__remote_consumer_groups = remote_consumer_groups
+        self.__get_commit_log_streams_for_topic = get_commit_log_streams_for_topic
+
+        self.__lock = Lock()
+        self.__subscription: Subscription[TStream, TOffset] = Subscription([])
+        self.__shutdown_requested = Event()
+
+        # TODO: If the commit log consumer crashes, it should cause all method
+        # calls on this object to raise (probably a RuntimeError raised from
+        # the commit log consumer exception.)
+        self.__commit_log_consumer_future = execute(self.__run_commit_log_consumer)
+
+    def __run_commit_log_consumer(self) -> None:
+        # TODO: How does the commit log consumer ensure that it is the only
+        # consumer within it's consumer group (or that it otherwise has the
+        # comprehensive set of partitions assigned for the commit log topic?)
+        # TODO: How do we ensure that the commit log consumer is not committing
+        # it's offsets (or is subject to automatic offset resets?)
```

Comment:
Right now these are properties of the Kafka configuration, which is private to that class. I've been mulling over if these configuration values should be handled differently, and I think they should, but I haven't come up with a good solution yet.

Broadly speaking, I think it'd be helpful for things like the group ID to be passed down to the balanced consumer constructor. Maybe other configuration values that are common to all backends should be encapsulated in a configuration class?